bindings/python/tests/test_transaction.py

```diff
@@ -0,0 +1,73 @@
+from pathlib import Path
+
+try:
+    import pytest
+except ImportError:
+    HAVE_PYTEST = False
+else:
+    HAVE_PYTEST = True
+
+import dlite
+
+
+# Configure paths
+thisdir = Path(__file__).parent.absolute()
+
+#dlite.storage_path.append(thisdir / 'entities' / '*.json')
+dlite.storage_path.append(thisdir / '*.json')
+Person = dlite.get_instance('http://onto-ns.com/meta/0.1/Person')
+
+
+person = Person(dims={"N": 4})
+person.name = "Knud H. Thomsen"
+person.age = 30
+person.skills = ["writing", "humor", "history", "human knowledge"]
+
+# Update person 6 times and create a snapshot before each change
+for i in range(6):
+    person.snapshot()
+    person.age += 5
+
+assert person.age == 60
+assert person.get_snapshot(0).age == 60
+assert person.get_snapshot(1).age == 55
+assert person.get_snapshot(6).age == 30
+if HAVE_PYTEST:
+    with pytest.raises(dlite.DLiteError):
+        person.get_snapshot(7)
+
+assert person._get_parent_uuid() == person.get_snapshot(1).uuid
+assert person._get_parent_hash() == person.get_snapshot(1).get_hash()
+
+# Chech mutability
+assert person.is_frozen() == False
+assert person.get_snapshot(1).is_frozen() == True
+assert person.get_snapshot(2).is_frozen() == True
```

Comment:
Other than checking `is_frozen()` value, you might also want to check that it truly _is_ frozen? I.e., checking that the `is_frozen()` value is respected.

Reply:
Good point, we can add a few tests. But truly ensuring that the frozen flag is respected is not possible in C. All setter functions already checks for immutability, but if you have an array property, nothing prevents a user from getting a pointer to the array and change the memory it points to. 

In Python I have now added immutability checks to the `__setattr__()` and `__getattr__()` methods. `__getattr__()` checks if the returned object is a numpy array. If it is and the instance is frozen, the array is made read-only which will raise a ValueError if the user tries to assign to it.

Reply:
I simply meant to add a test to try and change values in the snapshots and check an exception is raised :) But great!